### ISSUE TITLE
Test plan for "local function attributes"

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#38801]
Championed issue: dotnet/csharplang#1888

- [x] Specification checked in to `csharplang`.
- [x] attribute on declaration
- [x] attributes on declaration are emitted
- [x] attribute on return type
- [x] attributes on return type are emitted
- [x] attribute on type parameters
- [x] attributes on type parameters are emitted
- [x] attribute on parameters
- [x] attributes on parameters are emitted
- [x] Correctness of `AttributeUsage`
  - [x] On declaration and return](https://github.com/dotnet/roslyn/blob/features/local-function-attributes/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs#L631)
  - [x] On parameters and type parameters
- [x] interaction with potentially ambiguous syntax?
- [ ] speculative semantic model #24135
- [x] semantic model on attribute arguments for method/return/param/type param attributes
- [x] ~~expected diagnostics when binding a local function statement in speculative semantic model~~
  - Speculative semantic models don't give diagnostics
- [x] Scrub references (skipped tests and such) referencing this issue
- [x] test `Obsolete` attribute (and related)
  - [x] Diagnostics about attributes occur in local function locations
- [x] test `Conditional` attribute
  - [x] test `Conditional` with bad target method (non-void returning)
- [x] test various nullability attributes
- [x] test `[EnumeratorCancellation]` attribute on parameter of async-iterator local function #40959
- [x] MethodImplAttribute
- [x] DllImportAttribute
- [x] `static extern` local functions
- [x] local functions with no body
- [x] ExcludeFromCodeCoverageAttribute #41129 
- [x] MarshalAsAttribute #41299
- [x] SkipLocalsInitAttribute #41183
- [x] test restricted attributes such as IsReadOnlyAttribute, IsUnmanaged, ExtensionAttribute, etc.
- [x] Caller...Attribute #41299
- [x] LanguageVersion (tests added in #38808)
  - [x] Including in parameters
- [x] Scoping inside full syntax (See LookupTests.cs)
  - [x] Scoping for attribute lists on incomplete statements https://github.com/dotnet/roslyn/pull/41379#issuecomment-585453870
- [x] Can't use locals inside attribute creation, e.g. calling other local functions, using captured variables
  - [x] Local constants should probably be valid
  - [x] Attribute argument with declaration pattern
  - [x] Attribute argument with call containing 'out var'
  - [x] Attribute constructor with `out` parameter (disallowed)
     - [x] Out to a local
     - [x] Out var
- [x] Referencing local function with ConditionalAttribute only in a "dead" condition `#if` should not produce unused warning
- [x] ConditionalAttribute should be disallowed on non-static local functions #41333

IDE:
- [x] Test interactive typing
- [x] Start with an attribute
- [x] Completion
- [x] Keyword recommendation (`extern`)

Before merge:

- [x] ConditionalAttribute disallowed on non-static local functions #41333
- [x] IDE interactive testing
- [x] [AttributeUsage](https://github.com/dotnet/roslyn/blob/features/local-function-attributes/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs#L631)
- [x] Prevent regressing EndToEndTests
- [x] Draft spec checked into proposals, awaiting validation by Neal

**Comment by RikkiGibson** [comment]
Lower priority: I found that the diagnostics for non-static, extern local functions are not as clear as they could be. It should probably say something like "an 'extern' local function must also be 'static'".

It makes me wonder if there's any point in requiring the 'static' modifier be present. But I did find that instance member methods are allowed to be 'extern', so it seemed like it made things more clear to require it.

```cs
class C
{
    void M()
    {
        // Local function 'local()' must declare a body because it is not marked 'static extern'.
        extern void local();
    }
}
```
